velox/exec/tests/MemoryCapExceededTest.cpp

```diff
@@ -0,0 +1,111 @@
+  // the last one that makes its way to the top 3 can change between runs.
+  std::string expectedErrorMsgRegex =
+      "Exceeded memory cap of 5.00MB when requesting 2.00MB. Task total: "
+      "5.00MB Peak: 5.00MB. Top 3 Operators \\(by aggregate usage across "
+      "all drivers\\): Aggregation_#2_x1: 1.77MB Peak: 4.00MB, "
+      "FilterProject_#1_x1: 12.00KB Peak: 1.00MB, .+_x1: 0B Peak: "
+      "0B. Failed Operator: Aggregation_#2: 1.77MB";
+  std::vector<RowVectorPtr> data;
+  for (auto i = 0; i < 100; ++i) {
+    data.push_back(makeRowVector({
+        makeFlatVector<int64_t>(
+            size, [&i](auto row) { return row + (i * 1000); }),
+        makeFlatVector<int64_t>(size, [](auto row) { return row + 3; }),
+    }));
+  }
+
+  // Plan created to allow multiple operators to show up in the top 3 memory
+  // usage list in the error message.
+  auto plan = PlanBuilder()
+                  .values(data)
+                  .project({"c0", "c0 + c1"})
+                  .singleAggregation({"c0"}, {"sum(p1)"})
+                  .orderBy({"c0"}, false)
+                  .planNode();
+  auto queryCtx = core::QueryCtx::createForTest();
+  queryCtx->pool()->setMemoryUsageTracker(
+      velox::memory::MemoryUsageTracker::create(
+          kMaxBytes, kMaxBytes, kMaxBytes));
+  CursorParameters params;
+  params.planNode = plan;
+  params.queryCtx = queryCtx;
+  try {
+    readCursor(params, [](Task*) {});
+    FAIL() << "Expected a MEM_CAP_EXCEEDED RuntimeException.";
+  } catch (const VeloxException& e) {
+    auto errorMessage = e.message();
+    ASSERT_THAT(errorMessage, ::testing::MatchesRegex(expectedErrorMsgRegex));
+  }
+}
+
+TEST_F(MemoryCapExceededTest, multipleDrivers) {
+  // Executes a plan that runs with ten drivers and query memory limit that
+  // forces it to throw MEM_CAP_EXCEEDED exception. Verifies that the error
+  // message contains information that acknowledges the existence of 10 drivers.
+  vector_size_t size = 1'024;
+  const int32_t numSplits = 100;
+  constexpr int64_t kMaxBytes = 12LL << 20; // 12MB
+  std::vector<RowVectorPtr> data;
+  for (auto i = 0; i < numSplits; ++i) {
+    auto rowVector = makeRowVector({
+        makeFlatVector<int32_t>(
+            size, [&i](auto row) { return row + (i * 1000); }),
+        makeFlatVector<int32_t>(size, [](auto row) { return row + 3; }),
+    });
+    data.push_back(rowVector);
+  }
+
+  auto plan = PlanBuilder()
+                  .values(data, true)
+                  .singleAggregation({"c0"}, {"sum(c1)"})
+                  .planNode();
+  auto queryCtx = core::QueryCtx::createForTest();
+  queryCtx->pool()->setMemoryUsageTracker(
+      velox::memory::MemoryUsageTracker::create(
+          kMaxBytes, kMaxBytes, kMaxBytes));
+
+  CursorParameters params;
+  params.planNode = plan;
+  params.queryCtx = queryCtx;
+  params.maxDrivers = 10;
+  VELOX_ASSERT_THROW(readCursor(params, [](Task*) {}), "x10");
```

Comment:
nit: consider checking more of the error message.

Reply:
unfortunately, the multiple driver case is pretty non deterministic with respect to which nodes make it to the top 3 and their memory usage, which is why I am only verifying a very specific part here and the rest is being verified in the the single driver test. please let me know if you have anything in specific that you would like to test here.

Reply:
Got it. Maybe add a small comment to explain this.

Reply:
done